### PR TITLE
[BUGFIX]: Check datatype of results before converting to DataFrame

### PR DESCRIPTION
**Pull request by marcusianlevine**
The [merged hotfix](https://github.com/apache/incubator-superset/pull/2412) #2412 resolved an [issue with Presto/Hive columns](https://github.com/apache/incubator-superset/issues/2398) #2398 which are lists of dictionaries.

However, this introduced issue #3934 with traditional RDB datasources like MS SQL Server, which only return lists of lists. Wrapping such results in an outer list causes pandas to interpret the list as a single column.

This PR attempts to resolve that issue by introspecting the datatypes of the columns in the first row of the returned `data`. The type check may not be checking for exactly the right thing, but we likely need some sort of conditional to avoid wrapping a list of lists in an outer list

### CODE REVIEW

**Comment by xrmx** [comment]
Could you please clean the branch from other people commits?

File: superset/sql_lab.py [hunk 1]
```diff
@@ -231,11 +231,12 @@ def handle_error(msg):
 
     # check whether the result set is comprised of lists or dict
     if data and len(data) > 0:
```
Comment (xrmx) [inline superset/sql_lab.py]: if data:

File: superset/sql_lab.py [hunk 2]
```diff
@@ -231,11 +231,12 @@ def handle_error(msg):
...
-        first_row_type = type(data[0])
-        if first_row_type is list or first_row_type is tuple:
-            df_data = np.array(data)
-        else:
+        first_row = data[0]
+        first_row_types = set([type(c) for c in first_row])
```
Comment (xrmx) [inline superset/sql_lab.py]: Is this more readable?
```
any([isinstance(c, dict) for c in first_row])
```

**Review by xrmx — CHANGES_REQUESTED** [review]
Better to add a test so we don't regress.

File: superset/sql_lab.py (hunk 1 above)
Comment (marcusianlevine) [reply superset/sql_lab.py]: Yep you're right the len check is redundant

File: superset/sql_lab.py (hunk 2 above)
Comment (marcusianlevine) [reply superset/sql_lab.py]: Yeah I just took a stab at refactoring for clarity

**Comment by marcusianlevine** [comment]
@xrmx how would you recommend testing this? AFAIK the only reason for using `list(data)` instead of `np.array(data)` would be when reading certain nested column types from Hive or Presto

The existing `sqllab_tests.py` has queries in it and those tests seem to be passing, but we only test against MySQL, SQLite, and Postgres from what I can tell.

Any suggestions how we might simulate an embedded dict column from the test suite?

**Comment by xrmx** [comment]
The easiest way to test that code would be to move it to an helper and unit test it.

**Comment by marcusianlevine** [comment]
Ok moved the df conversion logic to helper and wrote two simple unit tests.

Not sure the tests actually prevent regression because I'm mocking the query results...

**Comment by marcusianlevine** [comment]
@xrmx do you think this is ready to merge?

**Pull request merged**